fallback/s2a_fallback.go

```diff
@@ -0,0 +1,159 @@
+	}
+	return func(ctx context.Context, targetServer string, conn net.Conn, s2aErr error) (net.Conn, credentials.AuthInfo, error) {
+		fallbackTLSConfig := tls.Config{
+			MinVersion:         tls.VersionTLS13,
+			ClientSessionCache: nil,
+			NextProtos:         []string{alpnProtoStrH2},
+		}
+		fallbackDialer := &tls.Dialer{Config: &fallbackTLSConfig}
+		fbConn, fbErr := fallbackDialer.DialContext(ctx, "tcp", fallbackServerAddr)
+		if fbErr != nil {
+			grpclog.Infof("dialing to fallback server %s failed: %v", fallbackServerAddr, fbErr)
+			return nil, nil, fmt.Errorf("dialing to fallback server %s failed: %v; S2A client handshake with %s error: %w", fallbackServerAddr, fbErr, targetServer, s2aErr)
+		}
+
+		tc, success := fbConn.(*tls.Conn)
+		if !success {
+			grpclog.Infof("the connection with fallback server is expected to be tls but isn't")
+			return nil, nil, fmt.Errorf("the connection with fallback server is expected to be tls but isn't; S2A client handshake with %s error: %w", targetServer, s2aErr)
+		}
+
+		tlsInfo := credentials.TLSInfo{
+			State: tc.ConnectionState(),
+			CommonAuthInfo: credentials.CommonAuthInfo{
+				SecurityLevel: credentials.PrivacyAndIntegrity,
+			},
+		}
+		if grpclog.V(1) {
+			grpclog.Infof("ConnectionState.NegotiatedProtocol: %v", tc.ConnectionState().NegotiatedProtocol)
+			grpclog.Infof("ConnectionState.HandshakeComplete: %v", tc.ConnectionState().HandshakeComplete)
+			grpclog.Infof("ConnectionState.ServerName: %v", tc.ConnectionState().ServerName)
+		}
+		conn.Close()
+		return fbConn, tlsInfo, nil
+	}, nil
+}
+
+// DefaultFallbackDialerAndAddress returns a TLS dialer and a network address for it to dial with.
+// Example use:
+//
+//	    fallbackDialer, fallbackServerAddr := fallback.DefaultFallbackDialerAndAddress(fallbackAddr)
+//		dialTLSContext := s2a.NewS2aDialTLSContextFunc(&s2a.ClientOptions{
+//			S2AAddress:         s2aAddress, // required
+//			EnableV2:           true, // must be true
+//			FallbackOpts: &s2a.FallbackOptions{
+//				FallbackDialer: &s2a.FallbackDialer{
+//					Dialer:     fallbackDialer,
+//					ServerAddr: fallbackServerAddr,
+//				},
+//			},
+//	})
+//
+// The fallback server's certificate should be verifiable using OS root store.
+// The fallbackAddr is expected to be a network address, e.g. example.com:port. If port is not specified,
+// it uses default port 443.
+// In the returned function's TLS config, ClientSessionCache is explicitly set to nil to disable TLS resumption,
+// and min TLS version is set to 1.3.
+func DefaultFallbackDialerAndAddress(fallbackAddr string) (*tls.Dialer, string, error) {
+	fallbackServerAddr, err := processFallbackAddr(fallbackAddr)
+	if err != nil {
+		if grpclog.V(1) {
+			grpclog.Infof("error processing fallback address [%s]: %v", fallbackAddr, err)
+		}
+		return nil, "", err
+	}
+	fallbackTLSConfig := tls.Config{
+		MinVersion:         tls.VersionTLS13,
+		ClientSessionCache: nil,
+		NextProtos:         []string{alpnProtoStrHttp},
```

Comment:
Is there a reason we're only supporting HTTP/1.1 here?

Reply:
thanks Matt, I had the (likely outdated) thought that h2 is new, and 1.1 would have wider support...
anyway I tested again by only setting h2, it also works.
I'll add both here, so the client and server can pick one that's supported in both ends.

Reply:
If we want to add both, that's fine, but here we're only setting HTTP/1.1 and above we're only setting HTTP/2. I think we need both strings in both places.

Reply:
I just updated the code, 
for the fallback handshake function: I'm setting only `h2`, reason is that this is for GRPC and it speaks http2 only

for the fallback dialer function: which is for the http client use case. I'm setting both `h2` and `http 1.1`, where client and server can negotiate into one protocol they both like 

plz correct me if I misunderstood, thanks again!